src/main/java/io/aiven/kafka/connect/common/output/SinkSchemaBuilder.java

```diff
@@ -0,0 +1,183 @@
+
+import io.aiven.kafka.connect.common.config.OutputField;
+
+import io.confluent.connect.avro.AvroData;
+import org.apache.avro.Schema;
+import org.apache.avro.SchemaBuilder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/* Build schema for output fields based on SinkRecord values
+ * For multiple output fields schema looks like this:
+ * {
+ *    "type" "record", "fields": [
+ *       {"name": "key", type="RecordKeySchema"},
+ *       {"name": "offset", type="long"},
+ *       {"name": "timestamp", type="long"},
+ *       {"name": "headers", type="map"},
+ *       {"name": "value", type="RecordValueSchema"},
+ *    ]
+ * }
+ */
+public abstract class SinkSchemaBuilder {
+
+    private final Logger logger = LoggerFactory.getLogger(SinkSchemaBuilder.class);
+
+    private final Collection<OutputField> fields;
+
+    private final AvroData avroData;
+
+    private final boolean envelopeEnabled;
+
+    public SinkSchemaBuilder(final Collection<OutputField> fields,
+                      final AvroData avroData,
+                      final boolean envelopeEnabled) {
+        this.fields = fields;
+        this.avroData = avroData;
+        this.envelopeEnabled = envelopeEnabled;
+    }
+
+    public SinkSchemaBuilder(final Collection<OutputField> fields,
+                                final AvroData avroData) {
+        this.fields = fields;
+        this.avroData = avroData;
+        this.envelopeEnabled = true;
+    }
+
+    protected abstract String getNamespace();
+
+    public Schema buildSchema(final SinkRecord record) {
+        Objects.requireNonNull(record, "record");
+        if (Objects.isNull(record.keySchema())) {
+            throw new DataException("Record key without schema");
+        }
+        if (Objects.isNull(record.valueSchema())) {
+            throw new DataException("Record value without schema");
+        }
+        logger.debug("Create schema for record");
+        logger.debug("Record Key Schema {}", record.keySchema());
+        logger.debug("Record Value Schema {}", record.valueSchema());
+        return avroSchemaFor(record);
+    }
+
+    protected Schema avroSchemaFor(final SinkRecord record) {
+        if (envelopeEnabled) {
+            final SchemaBuilder.FieldAssembler<Schema> schemaFields =
+                SchemaBuilder
+                    .builder(getNamespace())
+                    .record("connector_records")
+                    .fields();
+            for (final OutputField f : fields) {
+                final Schema schema = outputFieldSchema(f, record);
+                schemaFields.name(f.getFieldType().name).type(schema).noDefault();
+            }
+            return schemaFields.endRecord();
+        } else {
+            return tryUnwrapEnvelope(record);
+        }
+    }
+
+    private Schema tryUnwrapEnvelope(final SinkRecord record) {
+        // envelope can be disabled only in case of single field
+        final OutputField field = getFields().iterator().next();
```

Comment:
Should we validate if it's only a single field? or others are ignored?

Also, `getFields()` and `fields` is used inconsistently on different methods
```suggestion
        // envelope can be disabled only in case of single field
        final OutputField field = fields.iterator().next();
```